### PR TITLE
ENH - Add modular Group Prox Newton solver

### PR DESCRIPTION
**Pull request by Badr-MOUFAD**
This implements a group proximal Newton solver to extend ``skglm``  to handle problems with non-quadratic datafit and group separable penalties.

It proceeds as follows:
- [x] implement ``GroupProxNewton``
- [x] unitest the solver for Group Logistic regression
- [x] perform benchmarks

------------
link to [benchmarks results](https://badr-moufad.github.io/share-benchmarks/grp_logreg/benchmark_group_logreg_benchopt_run_2022-10-19_13h22m36.html)
link to [benchmark repo](https://github.com/Badr-MOUFAD/benchmark_group_logreg)

### CODE REVIEW

**Comment by Badr-MOUFAD** [comment]
For the benchmarks,
I am not aware of any solver that solves Group Logistic regression problem
([yngvem](https://group-lasso.readthedocs.io/en/latest/api_reference.html#group_lasso.LogisticGroupLasso) do that but uses [cross entropy](https://github.com/yngvem/group-lasso/blob/aa1c77aa3ba8534d543e3eec6b6f522d7996e2bf/src/group_lasso/_group_lasso.py#L859-L860) as a loss).
Share with me If you have one in mind. I would like to add it to the benchmark.

**Comment by mathurinm** [comment]
@Badr-MOUFAD the cross entropy for 2 classes should be the logistic loss

File: skglm/solvers/group_prox_newton.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+from numpy.linalg import norm
+from skglm.solvers.base import BaseSolver
+from skglm.utils import check_group_compatible
+
+EPS_TOL = 0.3
+MAX_CD_ITER = 20
+MAX_BACKTRACK_ITER = 20
+
+
+class GroupProxNewton(BaseSolver):
```
Comment (mathurinm) [inline skglm/solvers/group_prox_newton.py:12]: thanks Badr, it looks really cool. Do you think there is hope to reduce code redundancy with the vanilla prox newton solver ?

Comment (Badr-MOUFAD) [reply skglm/solvers/group_prox_newton.py:12]: It's possible if we add ``fit_intercept`` to the solver.

**Comment by Badr-MOUFAD** [comment]
> Badr-MOUFAD the cross entropy for 2 classes should be the logistic loss

Pls have a look at [this potion the code](https://github.com/yngvem/group-lasso/blob/aa1c77aa3ba8534d543e3eec6b6f522d7996e2bf/src/group_lasso/_group_lasso.py#L859-L860). Also in the benchmark, we don't converge to the same objective.

**Comment by mathurinm** [comment]
https://en.wikipedia.org/wiki/Cross_entropy#Cross-entropy_loss_function_and_logistic_regression

no ?

File: skglm/solvers/group_prox_newton.py (hunk 1 above)
Comment (mathurinm) [reply skglm/solvers/group_prox_newton.py:12]: go ahead if it reduces the code duplication

**Comment by Badr-MOUFAD** [comment]
> https://en.wikipedia.org/wiki/Cross_entropy#Cross-entropy_loss_function_and_logistic_regression
> 
> no ?

This is what one would expect. Also, after checking, yngvem ``LogisticGroupLasso`` with groups of size one does not give similar results to ``scikit-learn``'s ``LogisiticRegression``.

I raised an issue https://github.com/yngvem/group-lasso/issues/36.

File: skglm/solvers/group_prox_newton.py [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+def _descent_direction(X, y, w_epoch, Xw_epoch, fit_intercept, grad_ws, datafit,
+                       penalty, ws, tol):
+    # given:
+    #   1) b = \nabla F(X w_epoch)
+    #   2) D = \nabla^2 F(X w_epoch)  <------>  raw_hess
+    # minimize quadratic approximation for delta_w = w - w_epoch:
+    #  b.T @ X @ delta_w + \
+    #  1/2 * delta_w.T @ (X.T @ D @ X) @ delta_w + penalty(w)
+    # In BCD, we leverage inequality:
+    #  penalty_g(w_g) + 1/2 ||delta_w_g||_H <= \
```
Comment (QB3) [inline skglm/solvers/group_prox_newton.py:156]: Should it be `||delta_w_g||_H^2` ?

File: skglm/solvers/group_prox_newton.py [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    #  penalty_g(w_g) + 1/2 * || H || * ||delta_w_g||
+    grp_ptr, grp_indices = penalty.grp_ptr, penalty.grp_indices
+    n_features_ws = sum([penalty.grp_ptr[g+1] - penalty.grp_ptr[g] for g in ws])
+    raw_hess = datafit.raw_hessian(y, Xw_epoch)
+
+    lipchitz = np.zeros(len(ws))
+    for idx, g in enumerate(ws):
+        grp_g_indices = grp_indices[grp_ptr[g]:grp_ptr[g+1]]
+        # norm(X[:, grp_g_indices].T @ np.diag(raw_hess) @ X[:, grp_g_indices], ord=2)
+        lipchitz[idx] = norm(_diag_times_X_g(
```
Comment (QB3) [inline skglm/solvers/group_prox_newton.py:166]: I guess you encapsulated the `_diag_times_X_g` into a function to isolate the jit, this might worth a comment

**Review by QB3 — COMMENTED** [review]
First round of review

File: skglm/solvers/group_prox_newton.py (hunk 2 above)
Comment (Badr-MOUFAD) [reply skglm/solvers/group_prox_newton.py:156]: Yes indeed, good catch!

File: skglm/solvers/group_prox_newton.py (hunk 3 above)
Comment (Badr-MOUFAD) [reply skglm/solvers/group_prox_newton.py:166]: I made a ``_diag_times_X_g`` function to avoid copying the columns of ``X``.
The same operation can be achieved
```python
norm(X[:, grp_g_indices] @ np.sqrt(raw_hess), ord=2)**2
```
I made a comment about that in the function definition
https://github.com/scikit-learn-contrib/skglm/blob/99482daf44739bf9337992a278067386185049a5/skglm/solvers/group_prox_newton.py#L350-L353

Comment (QB3) [reply skglm/solvers/group_prox_newton.py:167]: thx!

File: skglm/solvers/group_prox_newton.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+                       penalty, ws, tol):
+    # given:
+    #   1) b = \nabla F(X w_epoch)
+    #   2) D = \nabla^2 F(X w_epoch)  <------>  raw_hess
+    # minimize quadratic approximation for delta_w = w - w_epoch:
+    #  b.T @ X @ delta_w + \
+    #  1/2 * delta_w.T @ (X.T @ D @ X) @ delta_w + penalty(w)
+    # In BCD, we leverage inequality:
+    #  penalty_g(w_g) + 1/2 ||delta_w_g||^2_H <= \
+    #  penalty_g(w_g) + 1/2 * || H || * ||delta_w_g||^2
```
Comment (QB3) [inline skglm/solvers/group_prox_newton.py:157]: I guess this yields `|| H ||^2` now that there is a square l156

File: skglm/solvers/group_prox_newton.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+                X, raw_hess * X_delta_w_ws, grp_g_indices)
+
+            old_w_ws_g = w_ws[range_grp_g].copy()
+            stepsize = 1 / lipchitz[idx]
+
+            w_ws[range_grp_g] = penalty.prox_1group(
+                old_w_ws_g - stepsize * past_grads[range_grp_g], stepsize, g)
+
+            # X_delta_w_ws += X[:, grp_g_indices] @ (w_ws[range_grp_g] - old_w_ws_g)
+            _update_X_delta_w_ws(X, X_delta_w_ws, w_ws[range_grp_g], old_w_ws_g,
```
Comment (QB3) [inline skglm/solvers/group_prox_newton.py:202]: If possible I would add a comment as before to say it avoids to copy the cols of `X` (very smart btw)

File: skglm/solvers/group_prox_newton.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+    for idx, j in enumerate(grp_g_indices):
+        delta_w_j = w_ws_g[idx] - old_w_ws_g[idx]
+        if w_ws_g[idx] != old_w_ws_g[idx]:
+            X_delta_w_ws += delta_w_j * X[:, j]
+
+
+@njit
+def _X_g_T_dot_vec(X, vec, grp_g_indices):
+    # X[:, grp_g_indices].T @ vec
+    # but without copying the cols os X
```
Comment (QB3) [inline skglm/solvers/group_prox_newton.py:344]: typo: >> cols of X

**Review by QB3 — COMMENTED** [review]
Thanks a lot for the outstanding work @Badr-MOUFAD !
One last question, `GroupProxNewton` has not been added to all the tests in the `test_group.py` file, is there a specific reason?

**Comment by Badr-MOUFAD** [comment]
> One last question, GroupProxNewton has not been added to all the tests in the test_group.py file, is there a specific reason?

It's mainly to avoid redundancy and complications of  unittest:

1. for ``alpha_max`` test: group log reg and group lasso do not have the same ``alpha_max``
2. I didn't want to add tests for quadratic problems as prox newton is meant for non-quadratic ones, and we test it for log reg

**Review by QB3 — APPROVED** [review]
